### ISSUE TITLE
[Design]-ish Thoughts on Thunks 

### ISSUE THREAD

**Issue by Mornix** [JacquesCarette/Drasil#1631]
## Context
This issue relates to #1023 and #1164. 

### TL;DR
#1023: DocLang constructors are inconsistent. Less so than when #1023 was written, but we still have (many) constructors taking `[ConceptInstance]`, `[TheoryModel]`, `[a]` which are available in the `ChunkDB` as well. Then there's `Assumptions`, which is a constant constructor that is used as a form of token whose contents is pulled from the `ChunkDB` when needed.

#1164: DocLang is in a good place to extract additional information from it. Especially those sections of an SRS where it collects information from other places of the SRS. A lot of the infrastructure for this has been done with the use of Multiplate.

## State of Affairs
#1609 introduced Multiplate to DocLang as a means to do passes. Currently `Assumptions` is the only DocLang constructor which is “ideal” (in my opinion) because it reduces duplication of passed in information as well as captures (or doesn’t capture at the moment) display configuration as opposed to the actual content. Take the `IMs` constructor (as another example), the constructor is `[Sentence] -> Fields  -> [InstanceModel] -> DerivationDisplay -> SCSSub`, if we removed the duplicated information `[InstanceModel]`, then we’re left with a constructor that solely captures display information, something which `DocDesc` should exclusively capture. 

Looking at how `Assumptions` are retrieved:
https://github.com/JacquesCarette/Drasil/blob/23d2dfce3361a16abc9fdd64174dd303dca2e6ea/code/drasil-docLang/Drasil/DocumentLanguage.hs#L454-L456

This is the only place and time that expansion/retrieval is done. This leaves `Assumptions` as useless in other passes because `ChunkDB` isn’t passed in with `DocDesc` meaning we can’t perform introspection on `Assumptions`. A small sample of issues where this interaction occurs: #1628 and #1539.

As a sidenote — may play a factor when deciding a solution, currently `ChunkDB` is augmented with a pass over the DocLang which extracts chunks like `TheoryModel`s which get added to the DB. This mechanism currently leaks into example recipes which is inelegant. 

### The Simplest Solution
Really, since `Assumptions` is the only constructor which is a thunk, it could take `[ConceptInstance]` as an argument. 
* \+ (DocLang) Consistency
* \- Ease of Use (from recipes)

The simplest solution gets the job done, but is inelegant. It adds redundant data which should be elsewhere. 

### The Other Simplest Solution
Do nothing. This can be dealt with later.

### A Real Solution
This is fairly simple, but (in my opinion) an ineffective one.
Introduce a datatype:
```haskell
data Carrier a = Thunk
		| Expanded a
```
Where the `Assumptions` constructor would become
```haskell
Assumptions :: Carrier [ConceptInstance] -> SCSSub
```
Add a smart constructor to give us a thunk.
```haskell
mkAssumptions :: SCSSub
mkAssumptions = Assumptions Thunk
```
Finally expansion happens similar to how it does currently but through the use of [`traverseFor`](http://hackage.haskell.org/package/multiplate-0.0.3/docs/Data-Generics-Multiplate.html#v:traverseFor) and Multiplate.

This seems like a fairly reasonable solution. Why should we look beyond this? One of the limitations of this style of implementation is we don’t know if a `DocDesc` is in expanded form without looking at it. So type safety becomes an issue. What if we have a plate — or anything involving `DocDesc` — which wants to use the `[ConceptInstance]`s stored in `Assumptions` and we pass it some `Thunk`s instead? We end up with a runtime error of what is really a compile-time-known issue. Additionally, to (rightfully) satisfy the compiler, anytime we look at `Assumptions` we have to implement two pattern matches: one for `Thunk` and one for `Expanded`. This is tedious and awful.

* \+ Consistency
* \- Tedious
* \- Lack of Type Safety

### The Dream Solution: Types Are the Best — Types Are the Worst
The structure of a `DocDesc` remains the same before and after thunk expansion, however we (I) want to lift the notion of thunk expansion to the type level. Say as:
```haskell
data DocDesc (a :: * -> *) = …
```
Where `a` is either a `Proxy` (and thus a phantom type) or `Identity` where it is a carrier for some other type thus we define it as a kind. This resolves the problem of the last solution with pattern matching and runtime errors by lifting the notion of expansion to the type level.

Let’s scope this down to a concrete definition because another issue arises:
```haskell
data SCSSub (a :: * -> *) where
  Assumptions    :: a [ConceptInstance] -> SCSSub a
  TMs            :: [Sentence] -> Fields  -> [TheoryModel] -> SCSSub a
  GDs            :: [Sentence] -> Fields  -> [GenDefn] -> DerivationDisplay -> SCSSub a
  DDs            :: [Sentence] -> Fields  -> [DataDefinition] -> DerivationDisplay -> SCSSub a
  IMs            :: [Sentence] -> Fields  -> [InstanceModel] -> DerivationDisplay -> SCSSub a
  Constraints    :: Sentence -> Sentence -> Sentence -> [LabelledContent] -> SCSSub a
  CorrSolnPpties :: [Contents] -> SCSSub a
```
If we instantiate some structure of `SCSSub Proxy` and wish to populate it into an `SCSSub Identity` we would start like:
```haskell
expand :: SCSSub Proxy -> SCSSub Identity
expand (Assumptions Proxy) = Assumptions $ Identity $ …
```
But the rest of `SCSSub a` doesn’t use of care about `a`. The expansion for a `TMs` is the identity. However, to appease the type checker we must write
```haskell
expand (TMs s f lt) = TMs s f lt
```
And this becomes more tedious as we add more constructors or nested structures to `DocDesc`. 

Why not make a separate data type? Because really the DocLang is the same before and after, it merely has more information from “an external source”, for the convenience of later functions. Moreover it would be nice to abstract over the structure and make the changes where absolutely necessary as it reduces duplicated code. As far as I can tell `(DocDesc Proxy)✕ChunkDB` should be a bijection with `DocDesc Identity`. (Barring `ChunkDB` "junk" not used in the SRS)

What it seems like I want here is some functor-esque class return a (type correct) identity everywhere except for anywhere a `Proxy` is found and then carry out a specific action.

This sounds very close to what Multiplate does! 

Multiplate is a large fan of `Constant` and `Identity` functors, and what we want with `DocDesc` is _almost_ the `Identity`, but it doesn’t quite work. But that’s okay, it has [`traverseMFor`](http://hackage.haskell.org/package/multiplate-0.0.3/docs/Data-Generics-Multiplate.html#v:traverseMFor) where we can stage the transformation and after returning “run” it to retrieve a `DocDesc Identity`. However, making it into a monad and including a “run” would necessitate solving the issue coercing the trivial `SCSSub Proxy` to `SCSSub Identity` without explicit unpacking and repacking. Moreover, Multiplate’s typing doesn’t make it the most conducive to use on types which take a type parameter that morphs during use of a plate.

I think the largest outstanding issues with respect to the type-level thunks is the inability to automatically coerce phantom parameterized data constructors to another phantom type and creating a type-level definition (I think a type family could be enough for this) of what I think is pretty close to an `Either` which "satisfies" Multiplate.

* \+ Consistency
* \+ Type Safety
* ~~ Less Code?
* ~~Tedious?
* ~~Possible?

## Conclusion
There are a few different solutions to the DocLang consistency each with their own set of benefits and drawbacks. The ideal solution would be to lift thunks into the type-level, however it may result in very inconvenient code or not be feasible with the current facilities used/available in Drasil. There are ways to get the job done albeit it is inelegant. Or this could ultimately be a thing that is best not to delve into at the moment.

**Comment by Mornix** [comment]
#1668's merge completed what this issue was functionally aiming to implement. Closing.
